actors/evm/src/interpreter/precompiles/parameter.rs

```diff
@@ -18,212 +19,197 @@ impl From<GroupError> for PrecompileError {
     }
 }
 
-impl<'a> TryFrom<&'a [u8; 32]> for Parameter<u64> {
-    type Error = PrecompileError;
-
-    fn try_from(value: &'a [u8; 32]) -> Result<Self, Self::Error> {
-        assert_zero_bytes::<24>(value)?;
-        // Type ensures our remaining len == 8
-        Ok(Self(u64::from_be_bytes(value[24..].try_into().unwrap())))
+impl Parameter for u8 {
+    fn read(reader: &mut ParameterReader) -> Result<Self, PrecompileError> {
+        reader.drop_zeros::<31>()?;
+        Ok(reader.read_byte())
     }
 }
 
-impl<'a> TryFrom<&'a [u8; 32]> for Parameter<i64> {
-    type Error = PrecompileError;
-
-    fn try_from(value: &'a [u8; 32]) -> Result<Self, Self::Error> {
-        assert_zero_bytes::<24>(value)?;
-        // Type ensures our remaining len == 8
-        Ok(Self(i64::from_be_bytes(value[24..].try_into().unwrap())))
-    }
-}
-
-impl<'a> From<&'a [u8; 32]> for Parameter<U256> {
-    fn from(value: &'a [u8; 32]) -> Self {
-        Self(U256::from_big_endian(value))
-    }
+macro_rules! impl_param_int {
+    ($($t:ty)*) => {
+        $(
+            impl Parameter for $t {
+                fn read(reader: &mut ParameterReader) -> Result<Self, PrecompileError> {
+                    const ZEROS: usize = 32 - ((<$t>::BITS as usize) / 8);
+
+                    reader.drop_zeros::<ZEROS>()?;
+                    // Type ensures our remaining len
+                    Ok(<$t>::from_be_bytes(reader.read_fixed()))
+                }
+            }
+        )*
+    };
 }
 
-pub(super) type U256Reader<'a> = PaddedChunks<'a, u8, 32>;
+impl_param_int!(u16 i16 u32 i32 u64 i64);
 
-// will be nicer with https://github.com/rust-lang/rust/issues/74985
-/// Wrapper around `ChunksExact` that pads instead of overflowing.
-/// Also provides a nice API interface for reading Parameters from input
-pub(super) struct PaddedChunks<'a, T: Sized + Copy, const CHUNK_SIZE: usize> {
-    slice: &'a [T],
-    chunks: ChunksExact<'a, T>,
-    exhausted: bool,
+/// Provides a nice API interface for reading Parameters from input. This API treats the input as if
+/// it is followed by infinite zeros.
+pub(super) struct ParameterReader<'a> {
+    slice: &'a [u8],
 }
 
-impl<'a, T: Sized + Copy, const CHUNK_SIZE: usize> PaddedChunks<'a, T, CHUNK_SIZE> {
-    pub(super) fn new(slice: &'a [T]) -> Self {
-        Self { slice, chunks: slice.chunks_exact(CHUNK_SIZE), exhausted: false }
+impl<'a> ParameterReader<'a> {
+    pub(super) fn new(slice: &'a [u8]) -> Self {
+        ParameterReader { slice }
     }
 
-    pub fn next(&mut self) -> Option<&[T; CHUNK_SIZE]> {
-        self.chunks.next().map(|s| s.try_into().unwrap())
+    /// Drop a fixed number of bytes, and return an error if said bytes are not zeros.
+    pub fn drop_zeros<const S: usize>(&mut self) -> Result<(), PrecompileError> {
+        let split = S.min(self.slice.len());
+        let (a, b) = self.slice.split_at(split);
+        self.slice = b;
+        if a.iter().all(|&i| i == 0) {
```

Comment:
should be able to `if a == &[0; S][..split]` but thats an array access that we can avoid sooo... probably not?

Reply:
I think this way is technically faster. But it doesn't matter either way. I just wish there were a builtin version...